File: crates/graphforge-bindings-py/tests/slices.py
"""Real native Slice parity: boundary, frozen membership, revision and cancellation."""

import unittest
import uuid

import pyarrow as pa

import graphforge


def identity() -> str:
    return str(uuid.uuid4())


def encode(table: pa.Table) -> bytes:
    sink = pa.BufferOutputStream()
    with pa.ipc.new_stream(sink, table.schema) as writer:
        writer.write_table(table)
    return sink.getvalue().to_pybytes()


class SliceTests(unittest.TestCase):
    def test_native_slice_freeze_boundary_and_cancellation(self) -> None:
        graph = graphforge.GraphForge()
        graph.execute(
            "CREATE (a:Story {name: 'First'}), (b:Story {name: 'Second'}), "
            "(c:Character {name: 'Shared'}), (a)-[:FEATURES]->(c), (b)-[:FEATURES]->(c)"
        )
        request = {
            "request_uuid": identity(),
            "source": {"kind": "current"},
            "selector": {
                "kind": "filter",
                "label": "Character",
                "property": "name",
                "equals": "Shared",
            },
        }
        self.assertEqual(graph.preview_slice(request).num_rows, 1)
        self.assertEqual(graph.preview_slice(request, "boundary").num_rows, 4)
        version = identity()
        prepared = graph.prepare_research_version(
            {
                "operation_uuid": identity(),
                "version_uuid": version,
                "context_uuid": identity(),
                "label": None,
                "description": None,
                "created_at": 1,
                "required_versions": [],
            }
        )
        graph.commit_research_version_operation(prepared)
        request["source"] = {"kind": "version", "version_uuid": version}
        frozen = graph.freeze_slice(request)
        capsule = encode(frozen)
        graph.execute("CREATE (:Character {name: 'Later'})")
        self.assertEqual(graph.inspect_frozen_slice(capsule).num_rows, 1)
        old = graph.inspect_frozen_slice(capsule).column("object_uuid")[0].as_py()
        revised = graph.revise_frozen_slice(
            capsule,
            {"request_uuid": identity(), "exclude": {"nodes": [old]}, "source_version": None},
        )
        self.assertEqual(graph.inspect_frozen_slice(encode(revised)).num_rows, 0)
        token = graphforge.CancellationToken()
        token.cancel()
        with self.assertRaises(Exception) as cancelled:
            graph.preview_slice(request, cancellation=token)
        self.assertEqual(getattr(cancelled.exception, "code", None), "GF_CANCELLED")
        with self.assertRaises(Exception) as invalid:
            graph.freeze_slice({"private_sentinel": "private-secret-value"})
        self.assertNotIn("private_sentinel", str(invalid.exception))
        self.assertNotIn("private-secret-value", str(invalid.exception))
        graph.close()


def main() -> None:
    """Execute native Slice acceptance."""
    unittest.main()


if __name__ == "__main__":
    main()

File: crates/graphforge-cli/tests/slices.rs
//! Real CLI Arrow preview/freeze/revision over a durable Project.
use graphforge_api::*;
use std::{io::Cursor, path::Path, process::Command};
use uuid::Uuid;
fn run(root: &Path, args: &[&str]) -> std::process::Output {
    Command::new(env!("CARGO_BIN_EXE_gf"))
        .arg("--project")
        .arg(root)
        .args(["research", "slice"])
        .args(args)
        .output()
        .unwrap()
}
fn count(output: std::process::Output) -> usize {
    assert!(
        output.status.success(),
        "{}",
        String::from_utf8_lossy(&output.stderr)
    );
    arrow::ipc::reader::StreamReader::try_new(Cursor::new(output.stdout), None)
        .unwrap()
        .map(|b| b.unwrap().num_rows())
        .sum()
}
#[test]
fn cli_frozen_membership_survives_current_edits_and_reopen() {
    let directory = tempfile::tempdir().unwrap();
    let root = directory.path().join("project");
    let mut graph = GraphForge::new(root.to_str()).unwrap();
    graph
        .execute("CREATE (:Character {name:'Shared'})")
        .unwrap();
    let version = Uuid::now_v7();
    let operation = graph
        .prepare_research_version(PrepareResearchVersionRequest {
            operation_uuid: Uuid::now_v7(),
            version_uuid: version,
            context_uuid: Uuid::now_v7(),
            label: None,
            description: None,
            created_at: 1,
            required_versions: Default::default(),
        })
        .unwrap();
    graph
        .commit_research_version_operation(operation, &CancellationToken::new())
        .unwrap();
    drop(graph);
    let request = directory.path().join("request.json");
    let capsule = directory.path().join("slice.arrow");
    std::fs::write(&request, serde_json::to_vec(&serde_json::json!({"request_uuid":Uuid::now_v7(),"source":{"kind":"version","version_uuid":version},"selector":{"kind":"query","query":"MATCH (n:Character) RETURN n.node_uuid AS node_uuid"}})).unwrap()).unwrap();
    let file = request.to_str().unwrap();
    assert_eq!(count(run(&root, &["preview", "--file", file])), 1);
    let frozen = run(&root, &["freeze", "--file", file]);
    assert!(
        frozen.status.success(),
        "{}",
        String::from_utf8_lossy(&frozen.stderr)
    );
    std::fs::write(&capsule, frozen.stdout).unwrap();
    let graph = GraphForge::new(root.to_str()).unwrap();
    graph.execute("CREATE (:Character {name:'Later'})").unwrap();
    drop(graph);
    assert_eq!(
        count(run(
            &root,
            &["inspect", "--capsule", capsule.to_str().unwrap()]
        )),
        1
    );
    std::fs::write(
        &request,
        serde_json::to_vec(
            &serde_json::json!({"request_uuid":Uuid::now_v7(),"source_version":null}),
        )
        .unwrap(),
    )
    .unwrap();
    let revised = run(
        &root,
        &[
            "revise",
            "--capsule",
            capsule.to_str().unwrap(),
            "--file",
            file,
        ],
    );
    assert!(
        revised.status.success(),
        "{}",
        String::from_utf8_lossy(&revised.stderr)
    );
    std::fs::write(&capsule, revised.stdout).unwrap();
    assert_eq!(
        count(run(
            &root,
            &["inspect", "--capsule", capsule.to_str().unwrap()]
        )),
        1
    );
    std::fs::write(&request, br#"{"private_sentinel":"private-secret-value"}"#).unwrap();
    let invalid = run(&root, &["preview", "--file", file]);
    assert!(!invalid.status.success());
    let diagnostic = String::from_utf8_lossy(&invalid.stderr);
    assert!(!diagnostic.contains("private_sentinel"));
    assert!(!diagnostic.contains("private-secret-value"));
}
